src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+
+        private readonly int _length;
+        private readonly T[][] _items;
+
+        public SegmentedArray(int length)
+        {
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            if (length == 0)
+            {
+                _items = Array.Empty<T[]>();
+                _length = 0;
+            }
+            else
+            {
+                _items = new T[(length + s_segmentSize - 1) >> s_segmentShift][];
+                for (var i = 0; i < _items.Length - 1; i++)
+                {
+                    _items[i] = new T[s_segmentSize];
+                }
+
+                _items[^1] = new T[length & s_offsetMask];
+                _length = length;
+            }
+        }
+
+        private SegmentedArray(int length, T[][] items)
+        {
+            _length = length;
+            _items = items;
+        }
+
+        public bool IsFixedSize => true;
+
+        public bool IsReadOnly => false;
+
+        public bool IsSynchronized => false;
+
+        public int Length => _length;
+
+        public object SyncRoot => _items ?? Array.Empty<T[]>().SyncRoot;
+
+        public ref T this[int index]
+        {
+            get
+            {
+                return ref _items[index >> s_segmentShift][index & s_offsetMask];
+            }
+        }
+
+        int ICollection.Count => _length;
+
+        int ICollection<T>.Count => _length;
+
+        int IReadOnlyCollection<T>.Count => _length;
+
+        T IReadOnlyList<T>.this[int index] => this[index];
+
+        T IList<T>.this[int index]
+        {
+            get => this[index];
+            set => this[index] = value;
+        }
+
+        object IList.this[int index]
+        {
+            get => this[index];
+            set => this[index] = (T)value;
+        }
+
+        public object Clone()
+        {
+            var items = (T[][])_items.Clone();
+            for (var i = 0; i < items.Length; i++)
+            {
+                items[i] = (T[])items[i].Clone();
+            }
+
+            return new SegmentedArray<T>(_length, items);
+        }
+
+        public void CopyTo(Array array, int index)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                _items[i].CopyTo(array, index + (i * s_segmentSize));
+            }
+        }
+
+        void ICollection<T>.CopyTo(T[] array, int arrayIndex)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                ICollection<T> collection = _items[i];
+                collection.CopyTo(array, arrayIndex + (i * s_segmentSize));
+            }
+        }
+
+        public Enumerator GetEnumerator()
+            => new Enumerator(this);
+
+        int IList.Add(object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void ICollection<T>.Add(T value)
```

Comment:
so if Add isn't supported, i'm ont sure the expected use case for this collection type.  For example, with formatting, we don't know how many items we would need to support ahead of time, right?

Reply:
This type can be used to implement `SegmentedList<T>` the same way `T[]` can be used to implement `List<T>`.